File: 20221026/모범답안/20221026_09/yacht.js

/**
 * 
 * @param arr {number[]}
 * @return {number}
 */
function solution(arr) {
    let checker = new Checker(arr);
    return checker.getMaxScore();
}

class Checker {
    constructor(arr) {
        this.arr = arr;
        this.sum_arr = arr.reduce((acc, it) => acc + it, 0);
        this.sorted_arr = arr;
        this.sorted_arr = [...arr].sort((a, b) => a - b);

        this.diff = [];

        for (let i = 1; i < 5; i++) {
            this.diff.push(this.sorted_arr[i] - this.sorted_arr[i - 1]);
        }

        this.counts = [0, 0, 0, 0, 0, 0];
        this.arr.forEach(it => { this.counts[it - 1]++; });
        this.sorted_counts = [...this.counts].sort((a, b) => a - b);
    }

    checkYacht() {
        if (this.counts.find(it => it === 5) != undefined) {
            return 50;
        }

        return 0;
    }

    check4Kind() {
        if (this.sorted_counts[this.sorted_counts.length - 1] >= 4) {
            return this.sum_arr;
        }

        return 0;
    }

    checkFullHouse() {
        if (this.sorted_arr[0] === this.sorted_arr[1] && this.sorted_arr[2] === this.sorted_arr[4]) {
            return this.sum_arr;
        } else if (this.sorted_arr[0] === this.sorted_arr[2] && this.sorted_arr[3] === this.sorted_arr[4]) {
            return this.sum_arr;
        }

        return 0;
    }

    checkLStraight() {
        if (this.diff.every(it => it === 1)) {
            return 30;
        }

        return 0;
    }

    checkSStraight() {
        if (this.diff.slice(0, 3).every(it => it === 1)) {
            return 15;
        } else if (this.diff.slice(1, 4).every(it => it === 1)) {
            return 15;
        }
        
        return 0;    
    }

    checkOthers() {
        let score = 0;
        for (let i = 1; i <= 6; i++) {
            score = Math.max(score, i * this.counts[i - 1]);
        }

        return score;
    }

    getMaxScore() {
        return [
            this.checkYacht(),
            this.check4Kind(),
            this.checkFullHouse(),
            this.checkLStraight(),
            this.checkSStraight(),
            this.checkOthers()
        ].reduce((acc, it) => Math.max(acc, it), 0);
    }
}

File: 20221026/모범답안/20221026_09/yacht.py
class Checker:
    def __init__(self, arr) -> None:
        self.arr = arr
        self.sum_arr = sum(arr)
        self.sorted_arr = sorted(arr)
        self.diff = [self.sorted_arr[i] - self.sorted_arr[i - 1] for i in range(1, 5)]
        
        self.counts = [arr.count(i) for i in range(1, 7)]
        self.sorted_counts = sorted(self.counts)

    def checkYacht(self):
        if self.counts.count(5) != 0:
            return 50
        return 0

    def check4Kind(self):
        if self.sorted_counts[-1] >= 4:
            return self.sum_arr
        return 0

    def checkFullHouse(self):
        if self.sorted_arr[0] == self.sorted_arr[1] and self.sorted_arr[2] == self.sorted_arr[4]:
            return self.sum_arr
        elif self.sorted_arr[0] == self.sorted_arr[2] and self.sorted_arr[3] == self.sorted_arr[4]:
            return self.sum_arr
        else:
            return 0

    def checkLStraight(self):
        if self.diff.count(1) == 4:
            return 30
        return 0

    def checkSStraight(self):
        if self.diff[:3].count(1) == 3:
            return 15
        elif self.diff[1:].count(1) == 3:
            return 15
        return 0
    
    def checkOthers(self):
        score = 0
        for i in range(1, 7):
            score = max(score, i * self.counts[i - 1])
        
        return score
    
    def getMaxScore(self):
        return max(
            [
                self.checkYacht(),
                self.check4Kind(),
                self.checkFullHouse(),
                self.checkLStraight(),
                self.checkSStraight(),
                self.checkOthers()
            ]
        )

def solution(arr):
    '''
    :param arr: int[]
    :return: int
    '''

    checker = Checker(arr)
    return checker.getMaxScore()
